### ISSUE TITLE
readerwritercircularbuffer.h not included in latest release. Intentional? Why?

### ISSUE THREAD

**Issue by ddevienne** [cameron314/readerwriterqueue#120]
Although it appears to predate that release.
Is that an indication that the fixed-sized queue is not ready for prime-times somehow?
I'd like a queue as much for parallelism, as for potentially throttling the producer side to not accumulate too much memory in queue (producer fetches large data from the network, and consumer writes it to disk; or the reverse). That's a correct use-case for that variant, no?

Thanks for any insights on this. --DD

PS: I've successfully used it, and I'm very happy with the results, but would like confirmation it is under the same BSD license, and  ready for prime time.

PPS: Note that I already have Boost-licensed dependencies, and noticed your multi-thread queue is dual-licensed, but not this single-thread one. Since it's easier to get new dependencies accepted under existing licenses, I'd welcome a similar arrangement. FWIW.

**Comment by cameron314** [comment]
It wasn't merged into master back when the last release was done (bit hard to see in the history because the merge was just a fast-forward).

Yeah, that's seems like a reasonable use-case, although for a single-threaded case you can also achieve back-pressure with a simple "max size" threshold without needing the concurrent bits.

Yes, it's BSD-licensed. You can also use it under the terms of the Boost license (you can take this comment as explicit written permission). Please note, however, that the semaphore code upon which the blocking circular buffer depends has its own, separate license (zlib; see the comments in atomicops.h).

Since it contains no known bugs and it's been used in the wild for a few months now, so I think it's safe to release it. Stay tuned!

**Comment by cameron314** [comment]
There you go! https://github.com/cameron314/readerwriterqueue/releases/tag/v1.0.5

**Comment by ddevienne** [comment]
> for a single-threaded case you can also achieve back-pressure with a simple "max size" threshold

The producer and consumer are on separate threads, so I killed _two birds with one stone_ (although I'm not the bird killing kind), adding parallelism, and throttling, to avoid a faster producer to accumulate too much memory in the queue. (the number of items in the queue is a good proxy for memory in my case, with queue items having an upper-bound on memory used).

And all that thanks to your wonderful queue.

**Comment by ddevienne** [comment]
And thank you for the dual-license authorization, and the v1.0.5 release!

One last question here: What's the best way to further discuss use of your single-thread and multi-thread queues?
I noticed Discussions is not active on this repo. Shall I ask questions as GH Issues? On stack-overflow? (do you have a tag there?)
Example questions I'm considering:
- why the atomics headers? C++11 atomics not good enough?
- I can achieve fan-out of my pipeline with two single-thread queues from this repo. Fan-in should be using the multi-thread queues from your other repo?
- Do you have an equivalent fixe-sized multi-thread queue, to achieve throttling on fan-in as well?
- I currently use your queue on the client side, where 1-thread per-pipeline stage is OK. But I'd also like to use it server-side, with multiple client-connections, where a thread-pool is already used (via Boost.Asio). Do you see any way to combine your queues with thread-pools?
- your queues remind me a lot of Go Channels. Do you see any way to have the equivalent of the select statement in Go, that's waiting on several channels (a.k.a. your queus) and _wakes up_ when any of them is ready? For that fan-in case, with blocking fixed-sized queues, how can you wait on a dequeue from multiple queues basically?
- queue termination to end the consumer thread dequeuing from the queue, to end the blocking. I'm currently sending a sentinel value to detect that case, but again to make an analogy with Go Channel and select, you can close() a Go Channel, and that wakes up the waiting consumer (Go routine), which can know the wake-up is from a close, and not a dequeue. Having an explicit close() would be cleaner that using sentinels, no?

Your libraries open up so much for me, that my head is boiling with questions! :)
And I'd be keen to discuss the above with you (and the community) if possible.

**Comment by cameron314** [comment]
You can ask questions in Github issues. I'm on Stack Overflow but I don't monitor it for mentions of my data structures.

> why the atomics headers? C++11 atomics not good enough?

I wanted to support VS2010 which didn't have std::atomic support.

> I can achieve fan-out of my pipeline with two single-thread queues from this repo. Fan-in should be using the multi-thread queues from your other repo?

There's multiple ways to achieve fan-in. If you don't care about relative ordering, then my ConcurrentQueue could certainly do the job, though its internal design of many SPMC queues is the opposite of the more memory- (and possibly CPU-) efficient MPSC design that fits this use case.

> Do you have an equivalent fixe-sized multi-thread queue, to achieve throttling on fan-in as well?

Nope, sorry.

> I currently use your queue on the client side, where 1-thread per-pipeline stage is OK. But I'd also like to use it server-side, with multiple client-connections, where a thread-pool is already used (via Boost.Asio). Do you see any way to combine your queues with thread-pools?

Not sure I understand this question. If the threads in the threadpool are handling independent requests, there's not much that can be shared?

> Your queues remind me a lot of Go Channels. Do you see any way to have the equivalent of the select statement in Go, that's waiting on several channels (a.k.a. your queus) and wakes up when any of them is ready? For that fan-in case, with blocking fixed-sized queues, how can you wait on a dequeue from multiple queues basically?

You could do this with the BlockingConcurrentqueue, which basically waits on a semaphore internally. This is less efficient because it has to loop through all the internal sub-queues to find the one with an element, but it's forced by the particular design of the queue. Blocking on multiple OS-level handles is generally the way to go but is tricky to keep efficient in all cases.

> Queue termination to end the consumer thread dequeuing from the queue, to end the blocking. I'm currently sending a sentinel value to detect that case, but again to make an analogy with Go Channel and select, you can close() a Go Channel, and that wakes up the waiting consumer (Go routine), which can know the wake-up is from a close, and not a dequeue. Having an explicit close() would be cleaner that using sentinels, no?

Well, a channel is a higher-level primitive than a queue :-) It's possible to implement close() and have the blocking reader wake up and see the queue is closed... but then what? There's no way for the current API to propagate that status to the caller. And every call to enqueue/dequeue would need to first check if the queue is closed.

**Comment by ddevienne** [comment]
Thanks for all those answers.

Regarding atomics, any chance to have a define to force using the STD <atomic> and avoid that dependency?
Either explicitly, or implicitly based on __cplusplus's value? Feels wrong to _redo_ something as tricky as atomics when it's built-in to the platform (I'm on C++17 BTW, VS2019 and GCC 9.2).

For the thread-pool question, each request is handled by a pipeline as well (partial deserialization, decompression of the rest, deserialization, handling/processing per-se, then per-response (there can be several to a single request), serialization, compression, where each of those is a stage of the pipeline, and there can be several in-flight requests in that pipeline). That's to achieve concurrency within a client connection, as opposed to across client connections. One way to do that is via Boost.Asio strands, which are scheduled across the thread pool, and have internal queues. But for a pipeline, there's no guarantee an item will make progress across the pipeline, rather than the first queue accumulating all 1st stage items, then flooding the 2nd stage, etc... Hopefully that makes sense, and it is clearer now. I'm kinda thinking aloud here, since discovering your queues, how they could fit in the network server scenario, running on a thread pool.

**Comment by cameron314** [comment]
atomicops.h exposes an interface that is slightly different from std::atomic, so the queue would have to be rewritten. However, atomicops.h does use std::atomic where available in its implementation, so it should be safe and compliant with modern C++ versions (it's a shim wrapper, not a reimplementation). Note that atomicops.h also includes a lightweight semaphore implementation, which has no equivalent in base C++.

I'm still not entirely sure I understand your threadpool setup, but if you're not CPU-bound then asynchronous I/O tasks are definitely a nice way to increase throughput. Not sure how useful my queues would be in such a case.

**Comment by ddevienne** [comment]
Hi. Another set of questions :). Still related to the Go Channel analogy. In Go, it's recommended to use _Unbuffered Channels_, i.e. of depth 1, and thus the Channel is more of a synchronization primitive, than a queue proper. Would you consider it an abuse of your fixed queue to use it that way, as opposed to using an std::mutex + std::condition_variable, i.e. purely as an intuitive synchronization primitive, when using a depth of 1?

And talking about cond-var, I was wondering what makes the consumer dequeuing side thread wake-up in your queue, and how fast it does wake-up, compared to a cond-var?

Cond-vars can also have spurious wake-ups one needs to deal with; Does you queue suffer from that?
And if so, you handle that internally?

I sincerely hope all my questions are not bothering you. Just ignore me if they do I guess. Thanks, --DD

**Comment by cameron314** [comment]
I mean, you can use the queue that way if you'd like, but there's probably more efficient ways. I don't know Go, but from the docs, it seems like an unbuffered channel forces the sender to wait until the receiver is ready, so even a 1-slot queue wouldn't quite model this.

The blocking is accomplished using a semaphore (for both sender and receiver), so it's as fast as the underlying semaphore implementation, which at the OS level should be pretty much the same speed as waking up from any other blocking primitive.

On POSIX-y systems, the semaphore wrapper can be spuriously woken up by certain signals, but it handles this internally, so the queue itself should never run into a spurious wakeup.

I don't mind answering questions, but my response time can vary significantly :-)
